### PR TITLE
Vladislav Pozniak - 0

### PR DESCRIPTION
**Pull request by Rewerson**
## Name
Vladislav Pozniak

## Homework#
0

## Link to video with demo
https://img.rewerson.of.by/rubizza/rewerson_0.mp4

## Comment
All 3 levels implemented.

Usage:
ruby getdata.rb
ruby run.rb

First scripts parses page, collects needed files to data/ folder. Second script performs the task.

### CODE REVIEW

**Comment by taleh007** [comment]
@Rewerson you need to make hound happy :) Solve error and ping me

File: Vladislav Pozniak/0/run.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+table = Roo::Spreadsheet.open(current, extension: File.extname(current))
+
+print "What price are you looking for?\n>\s"
+item = gets.chomp
+
+found = 0
+products = []
+prices = []
+(9..table.last_row).each do |row|
+  next unless table.cell(row, 'A').to_s.start_with? item.upcase
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: Why `stat_with?`
Find substring

for example  'сардельки' will fail

File: Vladislav Pozniak/0/run.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+require 'roo-xls'
+
+# Level 1
+current = Dir.glob('./data/*').max
+table = Roo::Spreadsheet.open(current, extension: File.extname(current))
+
+print "What price are you looking for?\n>\s"
+item = gets.chomp
+
+found = 0
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: useless assignment

use Hash like
```Ruby
{
 product_name => price
}
```

that will make you data safe and pretty good. and you can replace `found` to `products.size`

File: Vladislav Pozniak/0/run.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+products = []
+prices = []
+(9..table.last_row).each do |row|
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]:
```suggestion
products = (9..table.last_row).each_with_object({}) do |row, products|
   name = table.cell(row, 'A')
      # is it correct product?
   price = table.cell(row, 7)
   products[name] = price
 end
```

File: Vladislav Pozniak/0/run.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  puts "'#{item}' can not be found in database."
+  exit
+end
+
+# Level 2
+prices_min = [].replace(prices)
+prices_max = [].replace(prices)
+month_min = File.basename(current, File.extname(current))
+month_max = File.basename(current, File.extname(current))
+(0...products.length).each do |prod|
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: u can use iteraaaator
:)

products.each do |name, price|

File: Vladislav Pozniak/0/run.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  Dir.glob('./data/*').each do |file|
+    sheet = Roo::Spreadsheet.open(file, extension: File.extname(file))
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: I hate loading 118 files for checking few rows

File: Vladislav Pozniak/0/run.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    (9..sheet.last_row).each do |row|
+      next unless sheet.cell(row, 'A').to_s == products[prod]
+
+      price = if File.basename(file, File.extname(file)) > '2017'
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: u have `file`

File: Vladislav Pozniak/0/run.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+# Level 3
+(0...prices.length).each do |price|
+  similar = []
+  found_similar = 0
+  (9..table.last_row).each do |row|
+    next if table.cell(row, 'A').to_s == products[price]
+    next if table.cell(row, 7).nil?
+
+    if (prices[price] - table.cell(row, 7)).abs <= 0.25
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: move DELTA in constant

File: Vladislav Pozniak/0/run.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
+require 'roo'
+require 'roo-xls'
+
+# Level 1
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: u can combine all 3 levels in one iterator and it will make me happy

File: Vladislav Pozniak/0/run.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  puts "Lowest for '#{products[prod].capitalize}' was on #{month_min} \
+at price #{prices_min[prod]} BYN"
+  puts "Maximum for '#{products[prod].capitalize}' was on #{month_max} \
+at price #{prices_max[prod]} BYN"
+end
+
+# Level 3
+(0...prices.length).each do |price|
+  similar = []
+  found_similar = 0
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: the same as in 1 level

File: Vladislav Pozniak/0/run.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        prices_min[prod] = price
+        month_min = File.basename(file, File.extname(file))
+      end
+      if prices_max[prod] < price
+        prices_max[prod] = price
+        month_max = File.basename(file, File.extname(file))
+      end
+    end
+  end
+  puts "Lowest for '#{products[prod].capitalize}' was on #{month_min} \
```
[the remark this answers is not in the document]
Comment (taleh007) [reply Vladislav Pozniak/0/run.rb]: or
```suggestion
  puts <<~LOWEST
    Lowest for '#{products[prod].capitalize}' was on #{month_min} at price #{prices_min[prod]} BYN
  LOWEST
```

File: Vladislav Pozniak/0/getdata.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+require 'open-uri'
+require 'roo'
+require 'roo-xls'
+require 'addressable/uri'
+require 'uri'
+
+MONTHS = { 'январь': '01', 'февраль': '02', 'март': '03', 'апрель': '04',
+           'май': '05', 'июнь': '06', 'июль': '07', 'август': '08',
+           'сентябрь': '09', 'октябрь': '10', 'ноябрь': '11', 'декабрь': '12' }
+MONTHS.freeze
```
Comment (taleh007) [inline Vladislav Pozniak/0/getdata.rb]: nope. 
```
MONTHS = { blalblabla }.freeze
```
This is the same action, but less lines

File: Vladislav Pozniak/0/getdata.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+PAGE = Nokogiri::HTML(open('http://www.belstat.gov.by/ofitsialnaya-statistika/makroekonomika-i-okruzhayushchaya-sreda/tseny/operativnaya-informatsiya_4/srednie-tseny-na-potrebitelskie-tovary-i-uslugi-po-respublike-belarus/'))
```
Comment (taleh007) [inline Vladislav Pozniak/0/getdata.rb]: Move link to constant (string), not Nokogiri object

File: Vladislav Pozniak/0/getdata.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+PAGE.freeze
+links = PAGE.css('a[href*="xls"]').map { |a| a['href'] }
+
+Dir.mkdir('data') unless File.directory?('data')
+links.each do |link|
+  link.prepend('http://www.belstat.gov.by') unless link.start_with? 'http://'
+  link = Addressable::URI.escape(link) unless link.include? '%'
+  puts 'Processing ' + link
```
Comment (taleh007) [inline Vladislav Pozniak/0/getdata.rb]: interpolation

File: Vladislav Pozniak/0/getdata.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  file = link.split('/').last
+  IO.copy_stream(URI.open(link), file)
+  table = Roo::Spreadsheet.open(file, extension: File.extname(file))
+  year = table.cell(3, 'A').split(' ')[-2] # Malformed original string, need to parse it from the end
```
Comment (taleh007) [inline Vladislav Pozniak/0/getdata.rb]: very strange. user regex to extract year

Comment (taleh007) [reply Vladislav Pozniak/0/getdata.rb]: .scan(/\d{4}/) or similiar

File: Vladislav Pozniak/0/getdata.rb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+PAGE = Nokogiri::HTML(open('http://www.belstat.gov.by/ofitsialnaya-statistika/makroekonomika-i-okruzhayushchaya-sreda/tseny/operativnaya-informatsiya_4/srednie-tseny-na-potrebitelskie-tovary-i-uslugi-po-respublike-belarus/'))
+PAGE.freeze
+links = PAGE.css('a[href*="xls"]').map { |a| a['href'] }
+
+Dir.mkdir('data') unless File.directory?('data')
+links.each do |link|
+  link.prepend('http://www.belstat.gov.by') unless link.start_with? 'http://'
+  link = Addressable::URI.escape(link) unless link.include? '%'
+  puts 'Processing ' + link
+  file = link.split('/').last
```
Comment (taleh007) [inline Vladislav Pozniak/0/getdata.rb]: use correct names -- `file_name`

File: Vladislav Pozniak/0/run.rb (hunk 5 above)
Comment (Rewerson) [reply Vladislav Pozniak/0/run.rb]: Shit happens :)

File: Vladislav Pozniak/0/run.rb [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+end
+
+if products.size.zero?
+  puts "'#{item}' can not be found in database."
+  exit
+end
+
+# Level 2
+month_min = File.basename(current, File.extname(current))
+month_max = File.basename(current, File.extname(current))
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: u 
```suggestion
month_max = month_min
```

File: Vladislav Pozniak/0/run.rb [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    Lowest for '#{name.capitalize}' was on #{month_min} at price #{price_min} BYN
+  LOWEST
+  puts <<~MAXIMUM
+    Maximum for '#{name.capitalize}' was on #{month_max} \at price #{price_max} BYN
+  MAXIMUM
+end
+
+# Level 3
+DELTA = 0.25 # delta for searching for similar prices
+DELTA.freeze
```
Comment (taleh007) [inline Vladislav Pozniak/0/run.rb]: useless freeze for Float

**Pull request merged**